File: native/machine/machine/devices/audio/service_clock.cpp
#include "machine/devices/audio/service_clock.h"

#include "machine/devices/audio/command_fifo.h"
#include "machine/devices/audio/contracts.h"
#include "machine/devices/audio/slot_bank.h"
#include "machine/scheduler/budget.h"
#include "machine/scheduler/device.h"

#include <utility>

namespace bmsx {

ApuServiceClock::ApuServiceClock(DeviceScheduler& scheduler, const ApuCommandFifo& commandFifo, const ApuSlotBank& slots)
	: m_scheduler(scheduler)
	, m_commandFifo(commandFifo)
	, m_slots(slots)
	, m_cpuHz(APU_SAMPLE_RATE_HZ) {}

void ApuServiceClock::reset() {
	m_sampleCarry = 0;
	m_availableSamples = 0;
	m_scheduler.cancelDeviceService(DeviceServiceApu);
}

i64 ApuServiceClock::captureSampleCarry() const {
	return m_sampleCarry;
}

i64 ApuServiceClock::captureAvailableSamples() const {
	return m_availableSamples;
}

void ApuServiceClock::restore(i64 sampleCarry, i64 availableSamples) {
	m_sampleCarry = sampleCarry;
	m_availableSamples = availableSamples;
}

void ApuServiceClock::setCpuHz(i64 cpuHz) {
	m_cpuHz = cpuHz;
}

void ApuServiceClock::clearBudget() {
	m_sampleCarry = 0;
	m_availableSamples = 0;
}

void ApuServiceClock::accrueCycles(int cycles) {
	m_availableSamples += accrueBudgetUnits(m_cpuHz, APU_SAMPLE_RATE_HZ, m_sampleCarry, cycles);
}

bool ApuServiceClock::pendingSamples() const {
	return m_availableSamples != 0;
}

i64 ApuServiceClock::consumeSamples() {
	return std::exchange(m_availableSamples, 0);
}

void ApuServiceClock::scheduleNext(i64 nowCycles) {
	if (!m_commandFifo.empty()) {
		m_scheduler.scheduleDeviceService(DeviceServiceApu, nowCycles);
		return;
	}
	if (m_slots.activeMask() == 0u) {
		m_scheduler.cancelDeviceService(DeviceServiceApu);
		m_sampleCarry = 0;
		m_availableSamples = 0;
		return;
	}
	if (m_availableSamples > 0) {
		m_scheduler.scheduleDeviceService(DeviceServiceApu, nowCycles);
		return;
	}
	m_scheduler.scheduleDeviceService(DeviceServiceApu, nowCycles + cyclesUntilBudgetUnits(m_cpuHz, APU_SAMPLE_RATE_HZ, m_sampleCarry, 1));
}

} // namespace bmsx

File: machine/ts/src/machine/devices/audio/service_clock.ts
import { accrueBudgetUnits, cyclesUntilBudgetUnits, type BudgetAccrual } from '../../scheduler/budget';
import { DEVICE_SERVICE_APU, type DeviceScheduler } from '../../scheduler/device';
import type { ApuCommandFifo } from './command_fifo';
import { APU_SAMPLE_RATE_HZ } from './contracts';
import type { ApuSlotBank } from './slot_bank';

export class ApuServiceClock {
	private cpuHz = APU_SAMPLE_RATE_HZ;
	private sampleCarry = 0;
	private availableSamples = 0;
	private readonly budgetAccrual: BudgetAccrual = { wholeUnits: 0, carry: 0 };

	public constructor(
		private readonly scheduler: DeviceScheduler,
		private readonly commandFifo: ApuCommandFifo,
		private readonly slots: ApuSlotBank,
	) {}

	public reset(): void {
		this.sampleCarry = 0;
		this.availableSamples = 0;
		this.scheduler.cancelDeviceService(DEVICE_SERVICE_APU);
	}

	public captureSampleCarry(): number {
		return this.sampleCarry;
	}

	public captureAvailableSamples(): number {
		return this.availableSamples;
	}

	public restore(sampleCarry: number, availableSamples: number): void {
		this.sampleCarry = sampleCarry;
		this.availableSamples = availableSamples;
	}

	public setCpuHz(cpuHz: number): void {
		this.cpuHz = cpuHz;
	}

	public clearBudget(): void {
		this.sampleCarry = 0;
		this.availableSamples = 0;
	}

	public accrueCycles(cycles: number): void {
		accrueBudgetUnits(this.budgetAccrual, this.cpuHz, APU_SAMPLE_RATE_HZ, this.sampleCarry, cycles);
		this.sampleCarry = this.budgetAccrual.carry;
		this.availableSamples += this.budgetAccrual.wholeUnits;
	}

	public pendingSamples(): boolean {
		return this.availableSamples !== 0;
	}

	public consumeSamples(): number {
		const samples = this.availableSamples;
		this.availableSamples = 0;
		return samples;
	}

	public scheduleNext(nowCycles: number): void {
		if (!this.commandFifo.empty) {
			this.scheduler.scheduleDeviceService(DEVICE_SERVICE_APU, nowCycles);
			return;
		}
		if (this.slots.activeMask === 0) {
			this.scheduler.cancelDeviceService(DEVICE_SERVICE_APU);
			this.sampleCarry = 0;
			this.availableSamples = 0;
			return;
		}
		if (this.availableSamples > 0) {
			this.scheduler.scheduleDeviceService(DEVICE_SERVICE_APU, nowCycles);
			return;
		}
		this.scheduler.scheduleDeviceService(
			DEVICE_SERVICE_APU,
			nowCycles + cyclesUntilBudgetUnits(this.cpuHz, APU_SAMPLE_RATE_HZ, this.sampleCarry, 1),
		);
	}
}
